### ISSUE TITLE
Message reporting missing patterns for inexhaustive match is misleading

### ISSUE THREAD

**Issue by vandenheuvel** [rust-lang/rust#72377]
A small demo:
```rust
#[derive(PartialEq, Eq, Clone, Copy)]
enum X { A, B, C, }

fn main() {
  match (X::A, Some(X::A)) {
    (_, None) => false,
    (v, Some(w)) if v == w => true,
    (X::B, Some(X::C)) => false,
    (X::B, Some(X::A)) => false,
    (X::A, Some(X::C)) | (X::C, Some(X::A)) => false,
  };
}
```
[Playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=c73dc6c86512e340861643c9ce6284a6).

**I expected to see this happen**
The usual error message indicating that patterns `(A, Some(B))` and `(C, Some(B))` are not covered.

**Instead, this happened**
Patterns `(A, Some(A))` and `(A, Some(B))` are detected as not covered.

`rustc --version --verbose`:
```
rustc 1.45.0-nightly (a74d1862d 2020-05-14)
binary: rustc
commit-hash: a74d1862d4d87a56244958416fd05976c58ca1a8
commit-date: 2020-05-14
host: x86_64-unknown-linux-gnu
release: 1.45.0-nightly
LLVM version: 9.0
```

**Comment by jonas-schievink** [comment]
If you add the suggested arms, it will in fact detect the other ones as missing. It is expected that `(A, Some(A))` is detected as not covered since the match guard is not taken into account when checking exhaustiveness (and couldn't be, since it's an arbitrary expression).

**Comment by jonas-schievink** [comment]
Closing as expected behavior.

**Comment by vandenheuvel** [comment]
> It is expected that `(A, Some(A))` is detected as not covered since the match guard is not taken into account when checking exhaustiveness

Wouldn't you then expect that it mentions `(A, Some(A))`, `(B, Some(B))`, `(C, Some(C))`, `(A, Some(B))` and `(C, Some(B))` as not covered? Or is the number of outputs always restricted to two? In that case, I think it would be useful to indicate that some patterns might be missing, e.g.:
```
error[E0004]: non-exhaustive patterns: at least `(A, Some(A))` and `(A, Some(B))` not covered
```

**Comment by jonas-schievink** [comment]
Yeah that's true
